Reactor concurrent instances

nomisRev opened the pull request:
Adds `Mono` & `Flux` instances for the `Concurrent` typeclass.

aballano commented on arrow-fx-reactor/src/test/kotlin/arrow/fx/FluxKTest.kt:
```diff
@@ -164,26 +173,22 @@ class FluxKTest : UnitSpec() {
-    }
+  override fun FluxKOf<T>.eqv(b: FluxKOf<T>): Boolean {
+    val res1 = Try { value().timeout(Duration.ofSeconds(5)).blockFirst() }
+    val res2 = Try { b.value().timeout(Duration.ofSeconds(5)).blockFirst() }
```
I would avoid using Try as it'll only add more "weight" to removing it later

nomisRev replied:
This is not correct anyway.. I'm going to rewrite this `Eq` instance to take into account all values within the `Flux`.

nomisRev commented on arrow-fx-reactor/src/main/kotlin/arrow/fx/reactor/MonoK.kt:
```diff
@@ -225,31 +264,29 @@ data class MonoK<out A>(val mono: Mono<out A>) : MonoKOf<A> {
-        } catch (t: Throwable) {
-          cb(Left(t.nonFatalOrThrow()))
-          just(Unit)
+      Flux.create<A> { sink ->
```
Hehe. I was unable to implement this operator using `Mono` :/

It works fine through `Flux`. I'd opt to create a separate ticket for this, since it's an optimization rather than incorrect behavior.

nomisRev commented on arrow-fx-reactor/src/main/kotlin/arrow/fx/reactor/extensions/ForwardDisposable.kt:
```diff
@@ -0,0 +1,70 @@
+
+  private val state = atomic(init)
+
+  fun cancel(): Disposable {
```
It's actually a bit trickier than that 😅
This doesn't really cancel anything. `ForwardDisposable` exposes 2 functions.

- `cancel`, which returns a `Disposable` that can be linked to an `Sink` using `onCancel`.
- `complete`, to atomically set the `Disposable` of the stream that is running within the coupled `Sink`.

I kept this the same as for `IO` since it's only used internally and not exposed.

pakoito commented on arrow-fx-reactor/src/main/kotlin/arrow/fx/reactor/extensions/common.kt:
```diff
@@ -0,0 +1,7 @@
+import arrow.core.Tuple2
+import java.util.function.BiFunction
+
+fun <A, B> tupled(): BiFunction<A, B, Tuple2<A, B>> =
```
Scope it a bit so it doesn't appear on the global namespace. Internal should suffice IMO.

pakoito commented on arrow-fx-reactor/src/main/kotlin/arrow/fx/reactor/extensions/dispatchers.kt:
```diff
@@ -0,0 +1,36 @@
+fun Scheduler.asCoroutineContext(): CoroutineContext =
+  SchedulerContext(this)
+
+private class SchedulerContext(val scheduler: Scheduler) : AbstractCoroutineContextElement(ContinuationInterceptor), ContinuationInterceptor {
```
I thought we had this already someplace else.

nomisRev replied:
For RxJava.

pakoito commented on arrow-fx-reactor/src/test/kotlin/arrow/fx/FluxKTest.kt:
```diff
@@ -164,26 +173,22 @@ class FluxKTest : UnitSpec() {
-      }
-
-      errA == errB
-    }
```
Yikes

pakoito commented on arrow-fx-reactor/src/test/kotlin/arrow/fx/FluxKTest.kt:
```diff
@@ -164,26 +173,22 @@ class FluxKTest : UnitSpec() {
+        v1 == it
+      })
+    })
+  }
```
Yikes but a bit less

pakoito commented on arrow-fx-reactor/src/main/kotlin/arrow/fx/reactor/extensions/ForwardDisposable.kt:
```diff
@@ -0,0 +1,70 @@
+    when (current) {
+      is State.Active -> {
+        value.dispose()
+        throw IllegalStateException(current.toString())
```
Is this the semantics in `reactor`?

nomisRev replied:
Crutch

pakoito commented on arrow-fx-reactor/src/main/kotlin/arrow/fx/reactor/extensions/ForwardDisposable.kt:
```diff
@@ -0,0 +1,70 @@
+        value.dispose()
+        throw IllegalStateException(current.toString())
+      }
+      is State.Empty -> if (current == init) {
```
mmmmmmm `Disposable` does't have an identity. I'd use `===` to do reference comparison instead.

nomisRev replied:
Actually `EMPTY_DISPOSABLE` exists as an `empty` reference iirc.

pakoito commented on arrow-fx-reactor/src/main/kotlin/arrow/fx/reactor/extensions/ForwardDisposable.kt:
```diff
@@ -0,0 +1,70 @@
+ *
+ * This is used in `bracket`, `guaranteeCase` & `asyncF`.
+ */
+internal class ForwardDisposable {
```
Nit: interface and create instances in the companion object

nomisRev replied:
Okay

pakoito commented on arrow-fx-reactor/src/main/kotlin/arrow/fx/reactor/MonoK.kt:
```diff
@@ -97,33 +102,64 @@ data class MonoK<out A>(val mono: Mono<out A>) : MonoKOf<A> {
+
+  fun fork(coroutineContext: CoroutineContext): MonoK<Fiber<ForMonoK, A>> =
+    coroutineContext.asScheduler().let { scheduler ->
+      Mono.create<Fiber<ForMonoK, A>> { emitter ->
```
If the emitter is only used to return a value, it's better to use the blocking constructor. In `Rx` is something like `fromCompletable`.

nomisRev replied:
Right!

pakoito commented on arrow-fx-reactor/src/main/kotlin/arrow/fx/reactor/MonoK.kt:
```diff
@@ -97,33 +102,64 @@ data class MonoK<out A>(val mono: Mono<out A>) : MonoKOf<A> {
+          Mono.defer { f(ExitCase.Completed).value() }
+            .subscribe({ emitter.success() }, { e -> emitter.error(e) })
+        }))
+    }.k()
```
1. magic
2. pray
3. ???
4. cancellation

nomisRev replied:
Run this `Mono`, and guarantee calling `Cancelled`, `Completed` or `Error`.
Wire `Disposable` running this `Mono` to `Mono.create` using `ForwardDisposable`.

No magic or praying 😛

pakoito commented on arrow-fx-reactor/src/main/kotlin/arrow/fx/reactor/MonoK.kt:
```diff
@@ -97,33 +102,64 @@ data class MonoK<out A>(val mono: Mono<out A>) : MonoKOf<A> {
-    })
+        },
+        emitter::error,
+        { }
```
completion? do they need to be forwarded?

nomisRev replied:
No, the inner subcribed `Mono` calls the terminal operators.

pakoito commented on arrow-fx-reactor/src/main/kotlin/arrow/fx/reactor/FluxK.kt:
```diff
@@ -91,32 +93,45 @@ data class FluxK<out A>(val flux: Flux<out A>) : FluxKOf<A> {
-    })
+        },
+        { e: Throwable -> emitter.error(e) },
+        {}
```
Same as mono, does completion need to be forwarded?

nomisRev replied:
The inner one needs to call `completion` after this one completes...

Needs fixing.

pakoito commented on arrow-fx-reactor/src/main/kotlin/arrow/fx/reactor/FluxK.kt:
```diff
@@ -140,6 +155,18 @@ data class FluxK<out A>(val flux: Flux<out A>) : FluxKOf<A> {
+      Flux.create<Fiber<ForFluxK, A>> { emitter ->
+        val s: ReplayProcessor<A> = ReplayProcessor.create<A>()
+        val conn: reactor.core.Disposable = value()
+          .subscribeOn(scheduler)
```
Nit: The scheduler is only use here, the `let` may be too extra

pakoito commented on arrow-fx-reactor/src/main/kotlin/arrow/fx/reactor/MonoK.kt:
```diff
@@ -97,33 +102,64 @@ data class MonoK<out A>(val mono: Mono<out A>) : MonoKOf<A> {
+    }.k()
+
+  fun fork(coroutineContext: CoroutineContext): MonoK<Fiber<ForMonoK, A>> =
+    coroutineContext.asScheduler().let { scheduler ->
```
Nit: as in FluxK, this scheduler is only used in one place.

nomisRev commented:
@pakoito @aballano I removed `ForwardDisposable` since I was able to replace it with `Disposable.Composite` which also takes care of synchronization.

I addressed all comments besides the ones about `Eq`.
`Eq<ForFluxK>` is still incorrect atm, it should compare all values within the `Flux`.

pakoito commented on arrow-fx-reactor/src/main/kotlin/arrow/fx/reactor/FluxK.kt:
```diff
@@ -246,9 +248,9 @@ data class FluxK<out A>(val flux: Flux<out A>) : FluxKOf<A> {
-        val cancellable = ForwardDisposable()
-        sink.onCancel(cancellable.cancel())
-        cancellable.complete(
+        val cancellable = Disposables.composite()
```
Is this the one from reactor?

pakoito replied:
EDIT: yes.

pakoito commented on arrow-fx-reactor/src/main/kotlin/arrow/fx/reactor/extensions/fluxk.kt:
```diff
@@ -196,79 +197,107 @@ interface FluxKConcurrent : Concurrent<ForFluxK>, FluxKAsync {
         val dda = fa.value()
           .subscribeOn(scheduler)
           .subscribe(sa::onNext, sa::onError)
+          .also { forward.add(it) }
```
NIT: `also(forward::add)`

nomisRev replied:
That bailed with a compile error, `TypeVariable<(U)>`. Not sure what is going wrong there.

pakoito commented on arrow-fx-reactor/src/main/kotlin/arrow/fx/reactor/extensions/fluxk.kt:
```diff
@@ -196,79 +197,107 @@ interface FluxKConcurrent : Concurrent<ForFluxK>, FluxKAsync {
     other2.dispose()
     sink.error(err)
-  }
+  } else Unit
```
huh

I mean it's the most obvious one yet I'd never used it. I'd gone for some more clever yet crappier.

pakoito commented on arrow-fx/src/main/kotlin/arrow/fx/IO.kt:
```diff
@@ -1211,13 +1211,13 @@ fun <E, A> IOOf<E, A>.onException(finalizer: (Throwable) -> IOOf<E, Unit>): IO<E
-fun <E, A> IOOf<E, A>.fork(ctx: CoroutineContext = IO.dispatchers<Nothing>().default()): IO<E, Fiber<IOPartialOf<E>, A>> =
-  IO.async { cb ->
+fun <E, A> IOOf<E, A>.fork(ctx: CoroutineContext = IO.dispatchers<Nothing>().default()): IO<Nothing, Fiber<IOPartialOf<E>, A>> =
+  IO {
```
👌

pakoito commented:
Still approved.

nomisRev commented:
Thanks for the re-review Paco ❤️

aballano commented on arrow-fx-reactor/src/main/kotlin/arrow/fx/reactor/extensions/fluxk.kt:
```diff
@@ -157,6 +176,160 @@ interface FluxKAsync :
 
+interface FluxKConcurrent : Concurrent<ForFluxK>, FluxKAsync {
+
+  override fun <A> Kind<ForFluxK, A>.fork(coroutineContext: CoroutineContext): FluxK<Fiber<ForFluxK, A>> =
+    fix().fork(coroutineContext)
```
nit 

```suggestion
  override fun <A> Kind<ForFluxK, A>.fork(ctx: CoroutineContext): FluxK<Fiber<ForFluxK, A>> =
    fix().fork(ctx)
```

aballano commented on arrow-fx-reactor/src/main/kotlin/arrow/fx/reactor/extensions/fluxk.kt:
```diff
@@ -157,6 +176,160 @@ interface FluxKAsync :
+        .k()
+    }
+
+  private fun <A, B, C> onError(
+    active: AtomicBooleanW,
+    other2: ReactorDisposable,
+    other3: ReactorDisposable,
+    sink: FluxSink<RaceTriple<ForFluxK, A, B, C>>,
+    err: Throwable
+  ): Unit = if (active.getAndSet(false)) {
+    other2.dispose()
+    other3.dispose()
+    sink.error(err)
+  } else Unit
```
both onError functions could be local, on the other hand we could merge them into one since we don't really care about the parametric types:

```suggestion
  private inline fun onError(
    active: AtomicBooleanW,
    other2: ReactorDisposable,
    other3: ReactorDisposable?,
    sink: FluxSink<*>,
    err: Throwable
  ): Unit = if (active.getAndSet(false)) {
    other2.dispose()
    other3?.dispose()
    sink.error(err)
  } else Unit
```

(feel free to replace the 2nd disposable nullable type with something fancy if you prefer, a vararg would also be possible as the last argument)

aballano commented on arrow-fx-reactor/src/main/kotlin/arrow/fx/reactor/extensions/monok.kt:
```diff
@@ -118,6 +137,145 @@ interface MonoKAsync : Async<ForMonoK>, MonoKMonadDefer {
+    sink.error(err)
+  } else Unit
+
+  private fun <A, B, C> onError(
```
Same here, this could be grouped as one function

aballano commented on arrow-fx-reactor/src/main/kotlin/arrow/fx/reactor/FluxK.kt:
```diff
@@ -91,32 +94,48 @@ data class FluxK<out A>(val flux: Flux<out A>) : FluxKOf<A> {
+                Flux.defer { release(a, ExitCase.Error(e)).value() }
+                  .subscribe({ emitter.error(e) }, { e2 -> emitter.error(Platform.composeErrors(e, e2)) })
+              }, {
+                if (!acquireActive.value) emitter.complete()
```
The above meaning that we could potentially check this within a test if it makes sense :D

nomisRev replied:
Yes, you're right and it can be worse.

```
val acquire = Flux.just(10, 9, 8, 7, 6, 5, 4, 3, 2, 1).k()
val use = { i: Int -> Flux.timer(i, TimeUnit.SECONDS).map { i }.k() }
val release = { _, _ -> FluxK { println("Release $it") } }

acquire.bracketCase(release, use)
```

- All these timers will start running in parallel since we call `use` immediately in `onNext`
- The first element takes the longest to finish, so when `acquire` completes and `acquireActive` is set to false, and `Flux.timer(1, TimeUnit.SECONDS)` will finish before all other `use` calls are finished.

In other words the `emitter` is closed and the other values from `use` are never observed since they're dropped.

aballano replied:
Maybe instead of an AtomicBooleanW we can have an atomic of a list(?) and let onNext's push something (maybe just A) and onComplete push something different (Unit?), so when onNext's `use` finishes we pop onNext's own element and check for the one from onComplete in order to complete 🤔

rachelcarmena commented:
@nomisRev :raised_hands: the previous comment only shows `Concurrentqueue` because it's the only benchmark included:

```
jmh {
    include = [
            'arrow.benchmarks.Queue'
    ]
    ...
}
```

For instance, for this configuration:

```
jmh {
    include = [
            'arrow.benchmarks.Async',
            'arrow.benchmarks.Queue'
    ]
    ...
}
```

it would appear data for:

* Async.catsIO
* Async.io
* Async.scalazZIO
* Queue.concurrentQueue

**Default threshold** is being used. However, it can be changed in `compareBenchmarksCI` task configuration: https://47degrees.github.io/hood/gradle/ci/

nomisRev commented:
Closing due to getting deprecated.
